wakuv2/nwaku.go

```diff
@@ -2377,547 +2329,737 @@ func wakuNew(nodeKey *ecdsa.PrivateKey,
+	requestTypeDestroy
+	requestTypeStartDiscV5
+	requestTypeStopDiscV5
+	requestTypeVersion
+	requestTypeRelaySubscribe
+	requestTypeRelayUnsubscribe
+	requestTypePeerExchangeRequest
+	requestTypeConnect
+	requestTypeDialPeerByID
+	requestTypeListenAddresses
+	requestTypeENR
+	requestTypeListPeersInMesh
+	requestTypeGetConnectedPeers
+	requestTypeGetPeerIDsFromPeerStore
+	requestTypeGetPeerIDsByProtocol
+	requestTypeDisconnectPeerByID
+	requestTypeDnsDiscovery
+	requestTypeDialPeer
+	requestTypeGetNumConnectedRelayPeers
+)
 
-	if C.getRet(resp) == C.RET_OK {
-		return nil
-	}
-	errMsg := "error WakuStartDiscV5: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
+type request struct {
+	id         string
+	reqType    requestType
+	input      any
+	responseCh chan response
 }
 
-func (self *Waku) StopDiscV5() error {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuStopDiscV5(self.wakuCtx, resp)
+func NewWakuNode(ctx context.Context, config *WakuConfig) (*WakuNode, error) {
+	ctx, cancel := context.WithCancel(ctx)
 
-	if C.getRet(resp) == C.RET_OK {
-		return nil
+	n := &WakuNode{
+		newCh:  make(chan *request),
+		cancel: cancel,
 	}
-	errMsg := "error WakuStopDiscV5: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
-}
 
-func (self *Waku) WakuVersion() (string, error) {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
+	go func() {
+		defer gocommon.LogOnPanic()
 
-	C.cGoWakuVersion(self.wakuCtx, resp)
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
 
-	if C.getRet(resp) == C.RET_OK {
-		var version = C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-		return version, nil
-	}
+		C.waku_setup()
 
-	errMsg := "error WakuVersion: " +
-		C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return "", errors.New(errMsg)
-}
+		n.processLoop(ctx)
+	}()
 
-//export globalEventCallback
-func globalEventCallback(callerRet C.int, msg *C.char, len C.size_t, userData unsafe.Pointer) {
-	// This is shared among all Golang instances
-	self := Waku{wakuCtx: userData}
-	self.MyEventCallback(callerRet, msg, len)
+	_, err := n.postTask(requestTypeNew, config)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	return n, nil
 }
 
-func (self *Waku) MyEventCallback(callerRet C.int, msg *C.char, len C.size_t) {
-	fmt.Println("Event received:", C.GoStringN(msg, C.int(len)))
+func (n *WakuNode) postTask(reqType requestType, input any) (any, error) {
+	responseCh := make(chan response)
+	n.newCh <- &request{
+		reqType:    reqType,
+		input:      input,
+		responseCh: responseCh,
+	}
+	response := <-responseCh
+	if response.err != nil {
+		return nil, response.err
+	}
+	return response.value, nil
 }
 
-func (self *Waku) WakuSetEventCallback() {
-	// Notice that the events for self node are handled by the 'MyEventCallback' method
-	C.cGoWakuSetEventCallback(self.wakuCtx)
+func (n *WakuNode) processLoop(ctx context.Context) {
+
```

Comment:
I think it would be better to create both the request and response channel at this point and then avoid creating the response channel on every `postTask` call.

Another suggestion would be to handle all the req/resp channels logic in this `func` and make the `Waku`'s functions just return "ok" or "error" without dealing with channels